keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,200 @@
+
+KUDO currently has a feature that automatically restarts Pods of Deployments or StatefulSets if any KUDO-Parameter is changed. This
+is often required as a Parameter can change a value in a ConfigMap that is used by the Pods. If the Pods are not restarted
+the change is propagated to the Pods, but as most applications are not aware of changing config files they require a 
+restart. This can also be solved by other Controllers, for example [Reloader](https://github.com/stakater/Reloader).
+
+This behavior can currently not be controlled and is always active. The issue is that not all Parameter changes require
+a rolling restart of Pods, and sometimes the restart is unwanted and can negatively affect the state of the application
+the operator is controlling.
+
+One example would be the `replica` count of a StatefulSet: An increase here should only start new Pods and not restart
+all existing ones. Another example would be an application that regularly re-reads the config files, in this case the 
+pod doesn't need to be restarted if a config map changes - the application would pick up changed values automatically.
+
+The new flag will allow an operator developer to control when a changed parameter will trigger a restart of pods from 
+a deployment or stateful set.
+
+### Goals
+
+Make it possible for an operator developer to specify when a parameter will restart all pods from a stateful set or a deployment.
+
+### Non-Goals
+
+- Detailed control when Pods are restarted: If the spec template for a stateful set is changed, Kubernetes will still
+restart all pods automatically - there is no way to prevent that.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user explicitly disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+- Restarting a stuck Pod
+
+## Proposal
+
+Allow the operator developer to define groups of parameters that can then be used to trigger a restart of pods from
+a deployment or stateful set.
+
+Add a list of `groups` that each parameter is part of:
+
+```yaml
+  - name: CONFIG_VALUE
+    description: "A parameter that changes a config map which is used by a stateful set."
+    default: "3"
+    groups:
+      - MAIN_GROUP
+  - name: SOME_OTHER_PARAMETER
+    description: "A parameter that changes something that is used by two stateful sets"
+    default: "3"
+    groups:
+      - MAIN_GROUP
+      - ADDITIONAL_GROUP
+```
+
+KUDO then calculates the hash of all parameters in a group. This hash can then be used
+instead of the the `last-plan-execution-uid` in a deployment or stateful set to trigger the reload of the pods:
+
+```yaml
+spec:
+  template:
+    metadata:
+      annotations:
+        config-values-hash: {{ .ParamGroups.MAIN_GROUP.hash }}
```

Comment:
I'm not sure what the best syntax here would be. Maybe this one, or maybe

`.ParamGroups.Hashes.MAIN_GROUP`
or 
`.ParamGroupHashes.MAIN_GROUP`

If you have any other ideas, please let me know